core/src/main/java/com/linecorp/armeria/internal/common/stream/HttpDataDeframer.java

```diff
@@ -0,0 +1,417 @@
+                        throw new Error(); // Should not reach here.
+                }
+                remained -= readSize;
+                if (remained == 0) {
+                    break;
+                }
+            }
+
+            if (remained == 0) {
+                return value;
+            } else {
+                throw new IndexOutOfBoundsException(
+                        "readerIndex: " + (readerIndex() - 4 + remained) +
+                        " (expected: 0 <= readerIndex + 4 <= length(" + length() + "))");
+            }
+        }
+
+        /**
+         * Returns a newly retained slice of this {@link ByteBuf} at the current {@code readerIndex}
+         * and increases the {@code readerIndex} by the specified {@code length}.
+         *
+         * @throws IndexOutOfBoundsException if the specified {@code length} is greater than {@link #length()}
+         */
+        public ByteBuf readBytes(int length) {
+            // fast path
+            final ByteBuf firstBuf = queue.peek();
+            if (firstBuf.readableBytes() >= length) {
+                return firstBuf.readRetainedSlice(length);
+            }
+
+            // slow path
+            ByteBuf value = null;
+            int remained = length;
+            for (ByteBuf buf : queue) {
+                if (!buf.isReadable()) {
+                    continue;
+                }
+
+                final int readSize = Math.min(remained, buf.readableBytes());
+                if (readSize == length) {
+                    return buf.readRetainedSlice(readSize);
+                } else {
+                    if (value == null) {
+                        value = alloc.buffer(length);
+                    }
+                    value.writeBytes(buf, readSize);
+                    remained -= readSize;
+                    if (remained == 0) {
+                        break;
+                    }
+                }
+            }
+
+            if (remained > 0) {
+                ReferenceCountUtil.release(value);
+                throw new IndexOutOfBoundsException(
+                        "readerIndex: " + (readerIndex() - length + remained) +
+                        " (expected: 0 <= readerIndex + " + length + " <= length(" + length() + "))");
+            }
+
+            return value;
+        }
+
+        /**
+         * Returns a byte at the specified absolute {@code index} in this {@link HttpDataReader}.
+         * This method does not modify {@code readerIndex} of this {@link HttpDataReader}.
+         */
+        public byte getByte(int index) {
+            int remained = index;
+            for (ByteBuf buf : queue) {
+                final int length = buf.capacity();
+                if (length == 0) {
+                    continue;
+                } else if (length >= remained) {
+                    return buf.getByte(remained);
+                } else {
+                    remained -= length;
+                }
```

Comment:
Don't we need to take `buf.{reader,writer}Index()` into account unless we are sure all buffers in `queue` meets `readerIndex == 0 && writerIndex == capacity`?

Reply:
This block is refactored. 😀